gateway/rdp/iron_der.go

```diff
@@ -0,0 +1,468 @@
+				}
+			}
+			// multiple context-specific occurrences: build slice from each child
+			elemType := ft.Elem()
+			sliceVal := reflect.MakeSlice(ft, 0, len(children))
+			// special case: target is []byte
+			if elemType.Kind() == reflect.Uint8 {
+				// field is []byte; expect a single child representing an OCTET STRING.
+				if len(children) != 1 {
+					return errors.New("unexpected multiple elements for []byte field")
+				}
+				var b []byte
+				if _, err := asn1.Unmarshal(children[0].Bytes, &b); err != nil {
+					fieldVal.SetBytes(children[0].Bytes)
+					continue
+				}
+				// set bytes directly
+				fieldVal.SetBytes(b)
+				continue
+			}
+
+			// handle unsigned integer element types specially by decoding as int64 then converting
+			if elemType.Kind() >= reflect.Uint && elemType.Kind() <= reflect.Uint64 {
+				for _, ch := range children {
+					var vi int64
+					if _, err := asn1.Unmarshal(ch.Bytes, &vi); err != nil {
+						return err
+					}
+					elemPtr := reflect.New(elemType)
+					elemPtr.Elem().SetUint(uint64(vi))
+					sliceVal = reflect.Append(sliceVal, elemPtr.Elem())
+				}
+				fieldVal.Set(sliceVal)
+				continue
+			}
+
+			// handle slice elements that are structs by recursive unmarshal
+			if elemType.Kind() == reflect.Struct {
+				for _, ch := range children {
+					elemPtr := reflect.New(elemType)
+					if err := UnmarshalContextExplicit(ch.Bytes, elemPtr.Interface()); err != nil {
+						return err
+					}
+					sliceVal = reflect.Append(sliceVal, elemPtr.Elem())
+				}
+				fieldVal.Set(sliceVal)
+				continue
+			}
+
+			for _, ch := range children {
+				elemPtr := reflect.New(elemType)
+				if _, err := asn1.Unmarshal(ch.Bytes, elemPtr.Interface()); err != nil {
+					return err
+				}
+				sliceVal = reflect.Append(sliceVal, elemPtr.Elem())
+			}
+			fieldVal.Set(sliceVal)
+		default:
+			// direct type (int, uint64, string, []byte already handled above etc.)
+			// special-case unsigned integers: decode into int64 then convert
+			if ft.Kind() >= reflect.Uint && ft.Kind() <= reflect.Uint64 {
+				var vi int64
+				if _, err := asn1.Unmarshal(children[0].Bytes, &vi); err != nil {
+					return err
+				}
+				fieldVal.SetUint(uint64(vi))
+				continue
+			}
+			ptr := reflect.New(ft)
+			if _, err := asn1.Unmarshal(children[0].Bytes, ptr.Interface()); err != nil {
+				return err
+			}
+			fieldVal.Set(ptr.Elem())
+		}
+	}
+
+	return nil
+}
+
+// --- Marshal implementation -------------------------------------------------
+
+// MarshalContextExplicit encodes a struct (pointer or value) into DER where each
+// struct field with tag `asn1:"tag:N"` becomes a context-specific EXPLICIT
+// element [N] containing the normal DER encoding of the field value. The fields
+// are emitted in struct field order. Slices are encoded as a single SEQUENCE OF
+// unless the element is []byte (OCTET STRING), which is encoded as a single
+// OCTET STRING wrapper.
+func MarshalContextExplicit(v interface{}) ([]byte, error) {
```

Comment:
It seems this function is not used outside this package, make it private by renaming the function to `marshalContextExplicit`.

Is there any reason to not set the contents directly in a pointer known struct (*RDCleanPathPdu)?
From what I've seen, it will require less code to properly set the values into the proper type.

Try to avoid using the `reflect` library, you'll lose the type safety and it could lead to runtime panics.